### ISSUE TITLE
Always 11 to 13 dropped net_drops.eth0 : Is that relevant ?

### ISSUE THREAD

**Issue by UltimateByte** [firehol/netdata#988]
Hey there,

This is possibly a non-script issue, however, i think it's worth checking. I'd like to understand what's going on with either netdata and/or my machine.

One of my servers will always keep being between 11 and 13 dropped packets per hour, whatever the server load/bandwidth usage is.

![net_drops](http://image.noelshack.com/fichiers/2016/38/1474341406-12-packets-drops.png)

![net_drops_graph](http://image.noelshack.com/fichiers/2016/38/1474341406-packet-dropsped-inbound.png)

On the other hand, my ifconfig shows 2805 dropped packets : 

```
          RX packets:1887131167 errors:0 dropped:2805 overruns:0 frame:0
          TX packets:3609590062 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:255544211905 (237.9 GiB)  TX bytes:5314024262447 (4.8 TiB)
```

My questions are the following : Is this measurement relevant ?
How are those counted by netdata ?
How can i diagnose this then ? I'm open to any suggestion.

Thanks in advance for any help.

**Comment by ktsaou** [comment]
Netdata shows the incremental values. So your 2805 dropped packets will be 2816 - 2818 after an hour, 2827 - 2829 an hour after that, etc.

**Comment by UltimateByte** [comment]
Oh perfect. So those numbers matches, i'm glad to learn it.
Thanks for the answer.

Do you or anyone have an idea about what can be the cause of those drops ?

**Comment by ktsaou** [comment]
To my understanding dropped packets at an interface are always a problem. Packets are not just dropped for no reason.

I have encountered all kinds of possible reasons (from hardware interrupt issues, to wrong sizing of queues, to unsupported packets being sent by other other hosts, etc). In all cases though, there is is a fault somewhere.

So, I am also evaluating the situation. Should these alarms be hidden? At what level? More than 1, more than 10, more than 100? I don't know.

I have decided that it is better for netdata to ship with more alarms that you need and adapt them to your needs after you install it. This is why I have added the exact line and file these alarms are defined at their notifications. To make it easy for you see the alarm definition and decide if you need it or not.

Of course, I am still evaluating the situation. If it is a false alarm, it should be fixed. But I am not sure yet...

**Comment by UltimateByte** [comment]
Thank you for your answer.
I agree this probably matters.

You seem pretty well informed about those. Do you have any clue how i could diagnose the source of these drops ?
I'll also contact @cedarlug who is pretty well informed about network stuff, and will probably ask me to do some clever tcpdump :p

**Comment by ktsaou** [comment]
> You seem pretty well informed about those.

I am not that good. I am also researching solutions for my servers...

**Comment by UltimateByte** [comment]
Well, the only problem here is google didn't allow me to find very relevant stuff about packet drops... No well written articles, no research, just a few "i think that" things on forums. Must be pretty much the same for you then. I'll let you know if i find anything relevant about it.

**Comment by ktsaou** [comment]
On the netdata dashboard, at the system.softnet chart there is a reference to a redhat PDF that explains everything in detail...

**Comment by ktsaou** [comment]
This one: https://access.redhat.com/sites/default/files/attachments/20150325_network_performance_tuning.pdf

Red Hat has also published a lot of papers regarding Linux network fine tuning.

I am close this for now.

**Comment by UltimateByte** [comment]
Thanks for your attention. I'll try diagnosing drops and will feedback if i find anything relevant for Netdata and its users.

**Comment by ktsaou** [comment]
> I'll try diagnosing drops and will feedback if i find anything relevant for Netdata and its users.

perfect! thanks!

**Comment by UltimateByte** [comment]
OK, i think i finally found the source of those drops. They appear to occur around 35 seconds after a "milter-reject" : a mail trying to send out for a non-configured domain yet. It's trying to get out again and again in a periodic way, gets rejected, and causes a drop every single time. It's still unclear to me why it appears as a drop, probably because of how milter-reject works.

So in the end, this was a pain in the arse to find out, but the way i could diagnose it was pretty simple : 
1) Watch netdata Interface Drops (net_drops.eth0) in real time
2) Watch tail -f /var/log/syslog, in real time too
3) Try to correlate a syslog inpug and a drop

So i'll be configuring this domain's email. If it solves the issue, i won't report back as the is the expected behavior. If it doesn't, i will report back.

**Comment by ktsaou** [comment]
It is very strange for these packets to appear as network interface drops. interface drops are quite deep in the networking stack - these packets should have been dropped a lot higher...

But, let's see... my guess is that this won't stop the dropped packets...

**Comment by UltimateByte** [comment]
That's what i thought too. It makes no sense to me. Maybe milter-reject generates a bad checksum after all. Could be a bug or made on purpose.
Well, i'll report back in any case.. :)

**Comment by UltimateByte** [comment]
Wasn't about mail server...
Didn't figure out why it is happening yet.

**Comment by nbargnesi** [comment]
> Beginning with kernel 2.6.37, the rx_dropped counter shows statistics for dropped frames because of:
> Softnet backlog full
> Bad / Unintended VLAN tags
> Unknown / Unregistered protocols
>
>If any frames meet those conditions, they are dropped before the protocol stack and the rx_dropped counter is incremented. 

A bit more in-depth explanation from [Red Hat's knowledgebase is here](https://access.redhat.com/solutions/657483).

I've got two deployments of netdata showing non-zero `rx_dropped` ifconfig values where the same interfaces show no errors via `ethtool -S`.

tl;dr ifconfig RX dropped is not very useful - recommend reopening

**Comment by ktsaou** [comment]
Could you please post the output of the following script:

```sh
x="eth0"     # <<< set the interface here
echo "$x"; ifconfig $x; ethtool -S $x
echo "sysfs statistics"; for t in /sys/class/net/$x/statistics/{rx,tx}_*; do echo "$(basename $t): $(<$t)"; done
```

In general:

softnet backlog full is bad - there are softnet charts in netdata and separate alarms. Do they report dropped packets?

Bad VLANs and unregistered protocols do trigger dropped packets. The solution might be to increase the threshold. Do you believe this will solve it?

**Comment by nbargnesi** [comment]
Very low traffic on this particular instance (script output follows) - definitely not a softnet backlog issue. And no, no alarms for softnet specifically.

Increasing the threshold around VLANs and protocols wouldn't help since it doesn't address the core problem I don't think. Maybe the sum of all dropped values reported similar to how ethtool handles it?

Here's the script output:
```
eno2
eno2: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.0.0.4  netmask 255.255.255.0  broadcast 10.0.0.255
        ether 0c:c4:7a:33:6e:b1  txqueuelen 1000  (Ethernet)
        RX packets 29025  bytes 8325144 (7.9 MiB)
        RX errors 0  dropped 324  overruns 0  frame 0
        TX packets 8155  bytes 2919770 (2.7 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device memory 0xf7c00000-f7c7ffff  

NIC statistics:
     rx_packets: 29025
     tx_packets: 8155
     rx_bytes: 8441244
     tx_bytes: 2957974
     rx_broadcast: 1240
     tx_broadcast: 3375
     rx_multicast: 3141
     tx_multicast: 243
     multicast: 3141
     collisions: 0
     rx_crc_errors: 0
     rx_no_buffer_count: 0
     rx_missed_errors: 0
     tx_aborted_errors: 0
     tx_carrier_errors: 0
     tx_window_errors: 0
     tx_abort_late_coll: 0
     tx_deferred_ok: 0
     tx_single_coll_ok: 0
     tx_multi_coll_ok: 0
     tx_timeout_count: 0
     rx_long_length_errors: 0
     rx_short_length_errors: 0
     rx_align_errors: 0
     tx_tcp_seg_good: 386
     tx_tcp_seg_failed: 0
     rx_flow_control_xon: 0
     rx_flow_control_xoff: 0
     tx_flow_control_xon: 0
     tx_flow_control_xoff: 0
     rx_long_byte_count: 8441244
     tx_dma_out_of_sync: 0
     tx_smbus: 0
     rx_smbus: 0
     dropped_smbus: 0
     os2bmc_rx_by_bmc: 0
     os2bmc_tx_by_bmc: 0
     os2bmc_tx_by_host: 0
     os2bmc_rx_by_host: 0
     tx_hwtstamp_timeouts: 0
     rx_hwtstamp_cleared: 0
     rx_errors: 0
     tx_errors: 0
     tx_dropped: 0
     rx_length_errors: 0
     rx_over_errors: 0
     rx_frame_errors: 0
     rx_fifo_errors: 0
     tx_fifo_errors: 0
     tx_heartbeat_errors: 0
     tx_queue_0_packets: 343
     tx_queue_0_bytes: 19198
     tx_queue_0_restart: 0
     tx_queue_1_packets: 2140
     tx_queue_1_bytes: 1711143
     tx_queue_1_restart: 0
     tx_queue_2_packets: 3250
     tx_queue_2_bytes: 1024643
     tx_queue_2_restart: 0
     tx_queue_3_packets: 2422
     tx_queue_3_bytes: 164786
     tx_queue_3_restart: 0
     rx_queue_0_packets: 15267
     rx_queue_0_bytes: 4240575
     rx_queue_0_drops: 0
     rx_queue_0_csum_err: 0
     rx_queue_0_alloc_failed: 0
     rx_queue_1_packets: 5914
     rx_queue_1_bytes: 1851005
     rx_queue_1_drops: 0
     rx_queue_1_csum_err: 0
     rx_queue_1_alloc_failed: 0
     rx_queue_2_packets: 2386
     rx_queue_2_bytes: 721333
     rx_queue_2_drops: 0
     rx_queue_2_csum_err: 0
     rx_queue_2_alloc_failed: 0
     rx_queue_3_packets: 5458
     rx_queue_3_bytes: 1512231
     rx_queue_3_drops: 0
     rx_queue_3_csum_err: 0
     rx_queue_3_alloc_failed: 0
sysfs statistics
rx_bytes: 8325144
rx_compressed: 0
rx_crc_errors: 0
rx_dropped: 324
rx_errors: 0
rx_fifo_errors: 0
rx_frame_errors: 0
rx_length_errors: 0
rx_missed_errors: 0
rx_nohandler: 0
rx_over_errors: 0
rx_packets: 29025
tx_aborted_errors: 0
tx_bytes: 2919770
tx_carrier_errors: 0
tx_compressed: 0
tx_dropped: 0
tx_errors: 0
tx_fifo_errors: 0
tx_heartbeat_errors: 0
tx_packets: 8155
tx_window_errors: 0
```

**Comment by UltimateByte** [comment]
Here is my output http://hastebin.com/ekomekabup.rb

**Comment by ktsaou** [comment]
Interesting. In both outputs, the dropped packets reported by ifconfig are not shown by ethtool. Also the files in /sys/class/net seem to be the same with ifconfig ( @UltimateByte you didn't run the last command - so I don't know if that applies to your case too).

reading the source code of ethtool now...

**Comment by UltimateByte** [comment]
Sorry, read too fast, didn't see the script to run.
Here is the script output
http://hastebin.com/badikajulo.rb

**Comment by ktsaou** [comment]
thanks! same thing though. The /sys files report the exactly the same with ifconfig and /proc/net/dev.
It is unbelievable... the kernel is lying somehow... :)

The only solution I can think of is to write a special plugin that will make the same kernel calls `ethtool` does. I will try to research a bit more for alternatives.

**Comment by giacomofabbri** [comment]
Hi guys! I have the same problem, but actually It might not be a problem, idk! Anyway did you managed to solve this issue? Can I help you in some way gathering info? I'm not a linux expert so be patient with me! Thanks guys

**Comment by ktsaou** [comment]
Can you run the same commands:

```sh
x="eth0"     # <<< set the interface here
echo "$x"; ifconfig $x; ethtool -S $x
echo "sysfs statistics"; for t in /sys/class/net/$x/statistics/{rx,tx}_*; do echo "$(basename $t): $(<$t)"; done
```

Please post the output

**Comment by giacomofabbri** [comment]
After lunch I'm gonna try that, anyway just because I'm a noob, do I need to launch every single line separately or put everything inside a .sh script?? Thanks guys

**Comment by ktsaou** [comment]
Set the right interface name, and paste it on a terminal. It will do it.

**Comment by giacomofabbri** [comment]
That's the nethtool output

> ethtool -S eth0
NIC statistics:
     rx_packets: 3807331387
     tx_packets: 3577167934
     rx_bytes: 3492518105510
     tx_bytes: 4619279670967
     rx_broadcast: 66341
     tx_broadcast: 5558
     rx_multicast: 31691
     tx_multicast: 1141370
     rx_errors: 0
     tx_errors: 0
     tx_dropped: 0
     multicast: 31691
     collisions: 0
     rx_length_errors: 0
     rx_over_errors: 0
     rx_crc_errors: 0
     rx_frame_errors: 0
     rx_no_buffer_count: 527
     rx_missed_errors: 4954
     tx_aborted_errors: 0
     tx_carrier_errors: 0
     tx_fifo_errors: 0
     tx_heartbeat_errors: 0
     tx_window_errors: 0
     tx_abort_late_coll: 0
     tx_deferred_ok: 0
     tx_single_coll_ok: 0
     tx_multi_coll_ok: 0
     tx_timeout_count: 0
     tx_restart_queue: 979
     rx_long_length_errors: 0
     rx_short_length_errors: 0
     rx_align_errors: 0
     tx_tcp_seg_good: 802498938
     tx_tcp_seg_failed: 0
     rx_flow_control_xon: 0
     rx_flow_control_xoff: 0
     tx_flow_control_xon: 0
     tx_flow_control_xoff: 0
     rx_csum_offload_good: 3805974048
     rx_csum_offload_errors: 384
     rx_header_split: 0
     alloc_rx_buff_failed: 145
     tx_smbus: 0
     rx_smbus: 0
     dropped_smbus: 0
     rx_dma_failed: 0
     tx_dma_failed: 0
     rx_hwtstamp_cleared: 0
     uncorr_ecc_errors: 0
     corr_ecc_errors: 0
     tx_hwtstamp_timeouts: 0

The second command if I run as my user/sudo my user I get this

> sysfs statistics
-bash: /sys/class/net//statistics/rx_*: Permission denied
rx_*: 
-bash: /sys/class/net//statistics/tx_*: Permission denied
tx_*: 

If I run it as root instead

> sysfs statistics
bash: /sys/class/net//statistics/rx_*: No such file or directory
rx_*: 
bash: /sys/class/net//statistics/tx_*: No such file or directory
tx_*:

**Comment by giacomofabbri** [comment]
Actually I just run 
ethtool -S eth0
and then this 
echo "$x"; ifconfig $x; ethtool -S $xecho "sysfs statistics"; for t in /sys/class/net/$x/statistics/{rx,tx}_*; do echo "$(basename $t): $(<$t)"; done^C

So I may have done something wrong for sure

**Comment by giacomofabbri** [comment]
You may need also this 

> Link encap:Ethernet  HWaddr 4c:72:b9:66:c0:d4  
          inet addr:X.X.X.X  Bcast:X.X.X.X  Mask:255.255.255.0
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
          Interrupt:20 Memory:fe500000-fe520000

**Comment by ktsaou** [comment]
ok, could you please first become root and then run this line:

```sh
x="eth0"; echo "$x"; ifconfig $x; ethtool -S $x; echo "sysfs statistics"; for t in /sys/class/net/$x/statistics/{rx,tx}_*; do echo "$(basename $t): $(<$t)"; done
```

**Comment by ktsaou** [comment]
ok you have done it.
no need to do it.

let me check the output

**Comment by ktsaou** [comment]
I see that you have reset the counters of `ifconfig`: (zero bytes in/out and zero errors).

What I need to verify here is:

1. netdata shows dropped packets, that `ifconfig eth0` shows too.
2. the break-down of the network interface errors exposed by the kernel at `/sys/class/net/eth0/statistics is valid. I see that you searched for ` /sys/class/net//statistics/tx_`. Notice the `//`. There is no interface name there. So, you didn't run the command right. The commands I gave you need to be run on one terminal session, one after another. Do not innovate please.
3. the break-down `ethtool -S eth0` does match the output of `/sys/class/net/eth0/statistics`. (the other guys on this issue found a discrepancy: ethtool shows different output compared to `/sys/class/net/eth0/statistics`.

So, my request remains. Please run the one line I posted above and post its output.

**Comment by giacomofabbri** [comment]
Here's the correct output
`eth0
eth0      Link encap:Ethernet  HWaddr 4c:72:b9:66:c0:d4  
          inet addr:5.39.90.12  Bcast:5.39.90.255  Mask:255.255.255.0
          inet6 addr: fe80::4e72:b9ff:fe66:c0d4/64 Scope:Link
          inet6 addr: 2001:41d0:8:9c0c::/64 Scope:Global
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:3811442658 errors:0 dropped:7183 overruns:0 frame:0
          TX packets:3587842070 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:3492834266631 (3.4 TB)  TX bytes:4635258443813 (4.6 TB)
          Interrupt:20 Memory:fe500000-fe520000 

NIC statistics:
     rx_packets: 3811442771
     tx_packets: 3587842300
     rx_bytes: 3492834274089
     tx_bytes: 4635258789517
     rx_broadcast: 66577
     tx_broadcast: 5591
     rx_multicast: 31802
     tx_multicast: 1147278
     rx_errors: 0
     tx_errors: 0
     tx_dropped: 0
     multicast: 31802
     collisions: 0
     rx_length_errors: 0
     rx_over_errors: 0
     rx_crc_errors: 0
     rx_frame_errors: 0
     rx_no_buffer_count: 527
     rx_missed_errors: 4954
     tx_aborted_errors: 0
     tx_carrier_errors: 0
     tx_fifo_errors: 0
     tx_heartbeat_errors: 0
     tx_window_errors: 0
     tx_abort_late_coll: 0
     tx_deferred_ok: 0
     tx_single_coll_ok: 0
     tx_multi_coll_ok: 0
     tx_timeout_count: 0
     tx_restart_queue: 979
     rx_long_length_errors: 0
     rx_short_length_errors: 0
     rx_align_errors: 0
     tx_tcp_seg_good: 804363524
     tx_tcp_seg_failed: 0
     rx_flow_control_xon: 0
     rx_flow_control_xoff: 0
     tx_flow_control_xon: 0
     tx_flow_control_xoff: 0
     rx_csum_offload_good: 3810083724
     rx_csum_offload_errors: 385
     rx_header_split: 0
     alloc_rx_buff_failed: 145
     tx_smbus: 0
     rx_smbus: 0
     dropped_smbus: 0
     rx_dma_failed: 0
     tx_dma_failed: 0
     rx_hwtstamp_cleared: 0
     uncorr_ecc_errors: 0
     corr_ecc_errors: 0
     tx_hwtstamp_timeouts: 0
sysfs statistics
rx_bytes: 3492834277059
rx_compressed: 0
rx_crc_errors: 0
rx_dropped: 2229
rx_errors: 0
rx_fifo_errors: 0
rx_frame_errors: 0
rx_length_errors: 0
rx_missed_errors: 4954
rx_over_errors: 0
rx_packets: 3811442986
tx_aborted_errors: 0
tx_bytes: 4635259466157
tx_carrier_errors: 0
tx_compressed: 0
tx_dropped: 0
tx_errors: 0
tx_fifo_errors: 0
tx_heartbeat_errors: 0
tx_packets: 3587843121
tx_window_errors: 0`

Sorry for the trouble

**Comment by ktsaou** [comment]
well, you have several issues to solve:

```sh
# ethtool output
rx_no_buffer_count: 527
rx_missed_errors: 4954
rx_csum_offload_errors: 385
alloc_rx_buff_failed: 145
```

I suggest to search for answers. These are driver specific. Most probably you need to change a few settings at the driver.

I see that sysfs statistics has a discrepancy with ethtool. Although `rx_missed_errors` match, the other ones do not sum the `rx_dropped`.

```sh
# sysfs statistics
rx_dropped: 2229
rx_missed_errors: 4954
```

**Comment by ktsaou** [comment]
I see that at the dashboard at the dropped packet charts, we link to this document: https://www.novell.com/support/kb/doc.php?id=7007165

According to this document, dropped packets may mean:

1. Softnet backlog full (there is already another alarm for them at netdata)
2. Bad / Unintended VLAN tags
3. Unknown / Unregistered protocols
4. IPv6 frames when the server is not configured for IPv6

**Comment by hameno** [comment]
Yes, also: if you have bonded interfaces (active-backup mode) the slave interface counter increments with each ignored (broadcast) packet. This imho makes this counter practically useless for any monitoring. 
(Don't know why the kernel devs thought this would be a good change)
Could netdata switch to checking other values (ethtool)?

**Comment by ktsaou** [comment]
> Could netdata switch to checking other values (ethtool)?

Well, basic interface configuration can be obtained in a generic way using the `SIOCETHTOOL` `ioctl()`: http://stackoverflow.com/a/14266718

But, the statistics shown by ethtool are driver specific. So, we will need to support all possible drivers in netdata (which of course is a nightmare): http://git.kernel.org/cgit/network/ethtool/ethtool.git/tree/

The only generic breakdown of the dropped packets is `/sys/class/net/<INTERFACE>/statistics`, which as shown above, is not consistent with ethtool statistics.

**Comment by UltimateByte** [comment]
Just disabled IPv6 that was causing me trouble and rebooted. For now, no drops occurred.

**Comment by giacomofabbri** [comment]
I tried disabling IPv6 using this guide https://www.google.it/amp/www.techrepublic.com/google-amp/article/how-to-disable-ipv6-on-linux/?client=safari and rebooting but it doesn't seem to work unfortunately!!

**Comment by UltimateByte** [comment]
Also commented IPv6 lines in `/etc/network/interfaces`
Make sure IPv6 don't show up when you input: `ifconfig`
No more networking error since then.
https://webserver.terageek.org/netdata/

**Comment by ktsaou** [comment]
Cleaning up. I am closing this now, but if you believe there is something missing or problematic in netdata, just post to help us identify it and fix it.

**Comment by UltimateByte** [comment]
Now i get UDP errors when my bandwidth is saturated, so i assume this is normal. Also alarms don't last for too long now, which is good as well. :)
